Improve html parsing

hbostan opened the pull request:
Modified `articleParser` to add formatting to Telegram messages. It also shrinks down the image links by wrapping the name and size of the image in <a> tags. More over it replaces emojis in the html with their unicode values.

Added `emoji_codepoints.py`. It contains the byte values for emoji names

Added `HTMLSplitter`. It splits html into chunks with max size 4096. Generated chunks are also valid as html. This is needed to keep the formatting consistent across messages longer than 4096 characters.

hbostan committed:
Improvements to articleParser

- Added formatting support for telegram messages
- Improved article parser to handle image links better.
- Added emoji support

hbostan committed:
Remove old markdown parsers

Removed the parseLink function which matched links between html and
markdown. Also removed todo about markdown parsing and commented lines

hbostan committed:
Add HTMLSplitter

- Added HTMLSplitter to split long messages into 4096 byte chunks.
- Modified the bot's article sending logic for multiple chunks

hbostan committed:
html_splitter fixes

hbostan committed:
Emoji definition formatting

kadircet commented on newsparser.py:
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
+
+  # Tags supported by Telegram
+  supported_tags = [
+      'strike', 'i', 'strong', 'b', 'pre', 'ins', 'a', 'code', 'del', 's', 'em',
```
would be nice if you could put a link to some telegram bot api page

hbostan replied:
Done

kadircet commented on newsparser.py:
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
+  #               to provide (some) formatting in the telegram message, parser
+  #               converts tags that telegram doesn't like into ones it does
+  #               or removes them (not the contents) if no conversion exists.
+  # Handle Image: Images are wrapped in <a> tags with href same as img's src.
```
that's also nice, but it limits the message to have only a single image and lots of links. i think there's value in showing multiple images, for example as attachments.

no need to do it in this PR though, could you just put a todo and open an issue?

hbostan replied:
Added todo

kadircet commented on newsparser.py:
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
+
   def handle_starttag(self, tag, attrs):
-    if tag != "img":
+    tag_attrs = {at[0]: at[1] for at in attrs if at[0] in self.useful_attrs}
```
why not just create a mapping for all attributes, dropping the need for `useful_attrs`

hbostan replied:
done

kadircet commented on newsparser.py:
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
   def handle_starttag(self, tag, attrs):
-    if tag != "img":
+    tag_attrs = {at[0]: at[1] for at in attrs if at[0] in self.useful_attrs}
+    ptag = tag
```
can we do something like:
```
if tag == 'a':
  href = tag_attrs.get('href', '')
  if not href:
     log the tag and attr, as I suppose this should not happen
  self.append('<a href="{}">'.format(href)) # or whatever format string magic is there in python
  self.open_tags.append('a')
  return
if tag == 'img':
   check if it is an emoji, bail out if not
   also if it is not an emoji, check that last open_tag is 'a', and log the tag if not
   return
if tag not in supported_tags:
  tag = tag_mapping.get(tag)
if not tag:
  return
self.append(f'<{tag}>')
self.open_tags.append(tag)
```

hbostan replied:
done

kadircet commented on newsparser.py:
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
+    # it is never opened.
+    if tag not in self.supported_tags:
+      tag = self.tag_mapping.get(tag, None)
+    if tag != None:
```
```
if tag != self.open_tags[-1]:
  return
open_tags.pop
self.fed.append ...
```

hbostan replied:
done

kadircet commented on html_splitter.py:
```diff
@@ -0,0 +1,80 @@
+    return sum(len(item) for item in self.current_msg)
+
+  def getTotalMsgLen(self):
+    return self.getClosingTagsLen() + self.getCurrentMsgLen()
```
what about opening ?

kadircet commented on emoji_codepoints.py:
```diff
@@ -1,4147 +1,8288 @@
 # -*- coding: UTF-8 -*-
```
how is this file generated ?

can we instead check-in the generator and run it on the server at startup, to populate a data file mentioned in the config, if it doesn't exist or empty

hbostan committed:
Merge remote-tracking branch 'upsteam/master'

kadircet reviewed (approved):
As discussed offline let's get rid of the html splitter logic for now and keep the good old naive chunking.

hbostan committed:
Review changes

- Remove html_splitter and use legacy parser for long(>4096) messages.
- Refactor parser code
- Add emoji support and link cleanup to legacy parser
- Add emoji generator and remove emoji_codepoints file

hbostan committed:
add encoding to emoji_gen
